File: js/publish.js
const PubSub = require("@google-cloud/pubsub");


const listAllTopics = async () => {
  const pubsub = PubSub();
  // Lists all Pub/Sub topics in the current project
  return (await pubsub.getTopics())[0];
};

const createTopic = async (topicName) => {
  const pubsub = PubSub();
  // Creates a new Pub/Sub topic
  const topic = (await pubsub.createTopic(topicName))[0];
  console.log(`Topic ${topic.name} created.`);
  return topic;
};

const publishMessage = async (topicName, data) => {
  const allTopics = await listAllTopics();
  let nodeEnv = process.env.NODE_ENV;
  if (process.env.POD_NAMESPACE && process.env.POD_NAMESPACE !== "live") {
    nodeEnv = "beta";
  }
  let fullTopicName = `${topicName}-${nodeEnv}`;
  if (nodeEnv !== "production") {
    fullTopicName = `${fullTopicName}-${process.env.LOGNAME}-${process.env.TRAVIS_BRANCH}-${process.env.POD_NAMESPACE}`;
  }
  let topic = allTopics.find(t => t.name.split("/").slice(-1).pop() === fullTopicName);
  if (!topic) {
    topic = await createTopic(fullTopicName);
  }
  const publisher = topic.publisher();
  // Publishes the message as a string to the topic
  const dataBuffer = Buffer.from(JSON.stringify(data));
  try {
    const results = await publisher.publish(dataBuffer);
    const messageId = results[0];
    console.log(`Message ${messageId} published.`);
    return messageId;
  } catch (err) {
    console.log("Message not published.");
    console.error(err);
  }
};

module.exports = {
  listAllTopics: listAllTopics,
  createTopic: createTopic,
  publishMessage: publishMessage,
};

File: pubsub/publish.py
import os
import sys
import threading
import time
import logging

from google.cloud import pubsub_v1  # noqa E402

TESTING = sys.argv[1:2] == ['test']


def list_topics(project):
    """Lists all Pub/Sub topics in the given project."""
    publisher = pubsub_v1.PublisherClient()
    project_path = publisher.project_path(project)
    return publisher.list_topics(project_path)


def create_topic(project, topic_name):
    """Create a new Pub/Sub topic."""
    publisher = pubsub_v1.PublisherClient()
    topic_path = publisher.topic_path(project, topic_name)
    topic = publisher.create_topic(topic_path)
    logging.info('Topic created: {}'.format(topic))


def publish_message_async(topic_name, message):
    """Publishes multiple messages to a Pub/Sub topic."""
    if TESTING:
        return
    publisher = pubsub_v1.PublisherClient()
    project = os.getenv('GCLOUD_PROJECT', 'undefined')
    topics = list_topics(project)
    full_topic_name = topic_name + '-' + os.getenv('ENV', 'development')
    if os.getenv('ENV', 'development') != 'production':
        full_topic_name += ('-' + os.getenv('LOGNAME', 'undefined') + '-' +
                            os.getenv('TRAVIS_BRANCH', 'undefined') + '-' +
                            os.getenv('POD_NAMESPACE', 'undefined'))
    if len([topic for topic in topics if topic.name.split('/')[-1] == full_topic_name]) == 0:
        create_topic(project, full_topic_name)
    topic_path = publisher.topic_path(project, full_topic_name)
    data = str(message).encode('utf-8')
    try:
        publisher.publish(topic_path, data=data)
    except:
        time.sleep(60)
        publish_message(topic_name, message)


def publish_message(topic_name, message):
    msg_thread = threading.Thread(target=publish_message_async, args=(topic_name, message))
    msg_thread.start()
